zebra-chain/src/block/tests.rs

```diff
@@ -130,6 +136,160 @@ fn deserialize_block() {
+    };
+
+    // Serialize the block
+    let mut data = Vec::new();
+    block
+        .zcash_serialize(&mut data)
+        .expect("block should serialize as we are not limiting generation yet");
+
+    // Deserialize by now is ok as we are lower than the limit
+    let block2 = Block::zcash_deserialize(&data[..])
+        .expect("block should deserialize as we are just below limit");
+
+    assert_eq!(block, block2);
+    assert_eq!(max_transactions_in_block, block2.transactions.len());
+
+    // Add 1 more transaction to the block, limit will be reached
+    max_transactions_in_block = max_transactions_in_block + 1;
+    many_transactions.push(Arc::new(tx));
+    block.transactions = many_transactions.clone();
+    assert_eq!(max_transactions_in_block, many_transactions.len());
+
+    // Serialize will still be fine
+    let mut data = Vec::new();
+    block
+        .zcash_serialize(&mut data)
+        .expect("block should serialize as we are not limiting generation yet");
+
+    // Deserialize will now fail
+    Block::zcash_deserialize(&data[..]).expect_err("block should not deserialize");
+}
+
+#[test]
+fn block_limits_single_tx() {
+    // Test block limit with a big single transaction
+    // Dummy input and output
+    let input =
+        TransparentInput::zcash_deserialize(&zebra_test::vectors::DUMMY_INPUT1[..]).unwrap();
+    let output =
+        TransparentOutput::zcash_deserialize(&zebra_test::vectors::DUMMY_OUTPUT1[..]).unwrap();
+
+    // A block header
+    let blockheader = generate_block_header();
+
+    // Serialize header
+    let mut data_header = Vec::new();
+    blockheader
+        .zcash_serialize(&mut data_header)
+        .expect("Block header should serialize");
+
+    // Serialize a LockTime
+    let locktime = LockTime::Time(DateTime::<Utc>::from_utc(
+        NaiveDateTime::from_timestamp(61, 0),
+        Utc,
+    ));
+    let mut data_locktime = Vec::new();
+    locktime
+        .zcash_serialize(&mut data_locktime)
+        .expect("LockTime should serialize");
+
+    // Calculate the number of inputs we need
+    let mut max_inputs_in_tx = (MAX_BLOCK_BYTES as usize
+        - data_header.len()
+        - *&zebra_test::vectors::DUMMY_OUTPUT1[..].len()
+        - data_locktime.len())
+        / (*&zebra_test::vectors::DUMMY_INPUT1[..].len() - 1);
+
+    let mut outputs = Vec::new();
+
+    // Create inputs to be just below the limit
+    let mut inputs = std::iter::repeat(input.clone())
+        .take(max_inputs_in_tx)
+        .collect::<Vec<_>>();
+
+    // 1 single output
+    outputs.push(output);
+
+    // Create a big transaction
+    let big_transaction = Transaction::V1 {
+        inputs: inputs.clone(),
+        outputs: outputs.clone(),
+        lock_time: locktime,
+    };
+
+    // Put the big transaction into a block
+    let transactions = vec![Arc::new(big_transaction.clone())];
+    let block = Block {
+        header: blockheader,
+        transactions: transactions.clone(),
+    };
```

Comment:
Suggestion: Can you move this large block generation code to `large_single_transaction_block()`  in a `zebra_chain::test::generate` module?